UI plugin: Adds unamanaged cluster inventory page

christiang2 opened the pull request:
## What this PR does / why we need it
<!--
Add a detailed explanation of what this PR does and why it is needed.
-->
Adds unmanaged cluster landing page component
Adds unmanaged cluster reusable card component
Adds unmanaged cluster mock rest api endpoints

## Which issue(s) this PR fixes
<!--
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->
Fixes: #4654 

## Describe testing done for PR
<!--
Example: Created vSphere workload cluster to verify change.
-->
Populates list of unmanaged clusters from mock rest api

## Special notes for your reviewer
<!--
Add any things that reviewers should be aware of as they review
your PR.

Example: Please verify how I handled foo aligns with overall plan.
-->

<img width="1792" alt="Screen Shot 2022-06-02 at 6 59 05 PM" src="https://user-images.githubusercontent.com/106271458/171751862-c6965e9b-0749-4261-9114-67823c94453f.png">

Commit:
UI plugin: Adds unamanaged cluster inventory page

Adds unmanaged cluster landing page component
Adds unmanaged cluser reusable card component
Adds unmanaged cluster mock rest api endpoints

miclettej commented on cli/cmd/plugin/ui/web/tanzu-ui/src/shared/components/ClusterInfo/ClusterInfo.tsx:
```diff
@@ -0,0 +1,64 @@
+
+import './ClusterInfo.scss';
+
+function ClusterInfo(props: any) {
```
Since this is not a shared component across management, workload and unmanaged clusters I would rename it to something like UnmanagedClusterCard

miclettej commented on cli/cmd/plugin/ui/web/tanzu-ui/src/shared/components/ClusterInfo/ClusterInfo.scss:
```diff
@@ -0,0 +1,9 @@
+    color: grey;
+}
+
+cds-divider.app-custom {
```
I don't think this is being used anywhere, can remove

miclettej commented on cli/cmd/plugin/ui/web/tanzu-ui/src/shared/components/ClusterInfo/ClusterInfo.tsx:
```diff
@@ -0,0 +1,64 @@
+                    </div>
+                </div>
+                <div cds-layout="vertical gap:md align:vertical-center">
+                    <CdsDivider></CdsDivider>
```
If you move this divider outside of this div so it comes right after line 46, does the divider stretch to the full width of the card? That would follow the mockups a little more closely if you are able to make it work.

miclettej commented on cli/cmd/plugin/ui/web/tanzu-ui/src/views/unmanaged-cluster/UnmanagedClusterInventory.tsx:
```diff
@@ -1,34 +1,77 @@
+    const navigate = useNavigate();
+
+    useEffect(() => {
+        // fetch regions
```
comment should be 'fetch unmanaged clusters'

miclettej commented on cli/cmd/plugin/ui/web/tanzu-ui/src/views/unmanaged-cluster/wizard-basic/UnmanagedClusterBasic.tsx:
```diff
@@ -0,0 +1,24 @@
+// React imports
```
It is probably best to just remove this file from this pull request since it has a lot of references to AWS components and data store. It will make this PR cleaner but more importantly will make your commit to the main repo cleaner if you ever have to go back to this specific commit.

Commit:
Fixed styling to better match figma file
Removed unused code
Removed UnmangedClusterBasic no significant work done on it yet

miclettej commented on cli/cmd/plugin/ui/web/tanzu-ui/src/views/unmanaged-cluster/UnmanagedClusterInventory.tsx:
```diff
@@ -1,32 +1,72 @@
 // React imports
-import React from 'react';
+import React, { ChangeEvent, MouseEvent, useContext, useEffect, useState } from 'react';
```
I think ChangeEvent, MouseEvent, useContext are not used, can remove

Please also check other files that you added or modified for unused imports while you're at it.

miclettej commented on cli/cmd/plugin/ui/web/tanzu-ui/src/shared/components/UnmanagedClusterInfo/UnmanagedClusterInfo.tsx:
```diff
@@ -0,0 +1,64 @@
+                        </div>
+                        <CdsDivider orientation="vertical" cds-layout="align:right"></CdsDivider>
+                    </div>
+
```
nitpick: remove empty line

miclettej commented on cli/cmd/plugin/ui/web/tanzu-ui/src/shared/components/UnmanagedClusterInfo/UnmanagedClusterInfo.tsx:
```diff
@@ -0,0 +1,64 @@
+// App imports
+import './UnmanagedClusterInfo.scss';
+
+function UnmanagedClusterInfo(props: any) {
```
Very nice that you picked up on using props to pass data to a child component.

We can take this one step further. Instead of defining props as type 'any' what we should do right above this function is define an interface. The interface will define the 'shape' of the data that is expected when props are passed.

So above line 12 you will create that interface using 'export interface UnmanagedClusterInfoProps' and then define the attributes that are expected in that object (name, provider, status). 

Now what will happen is if someone goes to reuse this component, and they do not include required data that you define in that interface, the React compiler will throw an error to warn them.

Commit:
Removed empty spaces
Implemented interface for UnmanagedClusterInfoProps
Removed unused imports

swalner-vmware commented on cli/cmd/plugin/ui/web/tanzu-ui/node-server/src/routes/api/responses/unmanaged.json:
```diff
@@ -0,0 +1,13 @@
+    "provider": "Kind",
+    "status":"Offline"
+},{
+    "name": "UX/UI Cluster",
```
Might be nice to include Docker

christiang2 replied:
I think I was told the only two providers would be kind and Minikube, please let me know if this is inaccurate.

swalner-vmware commented on cli/cmd/plugin/ui/web/tanzu-ui/src/shared/components/UnmanagedClusterInfo/UnmanagedClusterInfo.scss:
```diff
@@ -0,0 +1,3 @@
+label {
+    color: grey;
```
I'd suggest re-using a Clarity grey, like in index.scss: .text-context-label {
    color: var(--cds-global-color-gray-700);

christiang2 replied:
Good point! Justin mentioned to use clarity options but it slipped my mind thank you!

swalner-vmware commented on cli/cmd/plugin/ui/web/tanzu-ui/src/shared/constants/NavRoutes.constants.tsx:
```diff
@@ -11,7 +11,7 @@ export const NavRoutes = {
     MANAGEMENT_CLUSTER_SELECT_PROVIDER: '/management-cluster-provider',
     WORKLOAD_CLUSTER_LANDING: '/workload-cluster-landing',
-
+    UMANAGED_CLUSTER_WIZARD: '/unmanaged-cluster-wizard',
```
I'm wondering if this route is more the `/unmanaged-cluster-landing` page (since it's just displaying the unmanaged clusters) and you'd want to reserve `/unmanaged-cluster-wizard` for the workflow that creates an unmanaged cluster?

christiang2 replied:
Yes that route will be for the cluster wizard I am currently in the process of developing it. Should I move the route? I put it there since I saw the workload cluster wizard there.

miclettej replied:
Yes this is correct to leave it as 'wizard' since this is for that wizard route

swalner-vmware commented on cli/cmd/plugin/ui/web/tanzu-ui/src/views/unmanaged-cluster/UnmanagedClusterInventory.tsx:
```diff
@@ -1,32 +1,72 @@
+                        Community Edition package repository is automatically installed when you deploy an unmanaged cluster.
+                    </div>
+                    <div cds-layout="vertical gap:md gap@md:lg col@sm:12 col:12">
+                        <CdsButton className="cluster-action-btn" status="primary" onClick={() => navigate(NavRoutes.DEPLOY_PROGRESS)}>
```
This is fine to me for now, but am I right in expecting that in the future this would link to the unmanaged wizard?

christiang2 replied:
Yes this is for the wizard but the wizard is not done so I have not linked it.

swalner-vmware commented on cli/cmd/plugin/ui/web/tanzu-ui/src/views/unmanaged-cluster/UnmanagedClusterInventory.tsx:
```diff
@@ -1,32 +1,72 @@
+                            onClick={() => {
+                                window.open('http://tanzucommunityedition.io', '_blank');
+                            }}
+                        >
```
I think we may want a deeper link to a page focused on unmanaged clusters

christiang2 replied:
Thank you for the feedback I think I was told that right section of the screen would be filled with more specific information in the future. But I do not have access to any of that information. If you have an ideas or know what goes there I would be happy to add it!

raymondz1 commented on cli/cmd/plugin/ui/web/tanzu-ui/src/shared/components/UnmanagedClusterInfo/UnmanagedClusterInfo.tsx:
```diff
@@ -0,0 +1,69 @@
+    status: string;
+}
+
+function UnmanagedClusterInfo(props: Partial<UnmanagedClusterInfoProps>) {
```
if we use `UnmanagedClusterInfoProps` instead of `Partial<UnmanagedClusterInfoProps>`, does it work?

christiang2 replied:
Yes it does, should I leave it as `UnmanagedClusterInfoProps`?

raymondz1 replied:
yes. please use UnmanagedClusterInfoProps instead of Partial<UnmanagedClusterInfoProps>

raymondz1 commented on cli/cmd/plugin/ui/web/tanzu-ui/src/views/unmanaged-cluster/UnmanagedClusterInventory.tsx:
```diff
@@ -1,32 +1,72 @@
+
+ClarityIcons.addIcons(disconnectIcon);
 
 const UnmanagedClusterInventory: React.FC = () => {
```
Please use function keyword for the function definition.

raymondz1 replied:
Oops. we missed that one. Could you please correct it in the PR? 
When we just started the project, we were using both ways to create components. But later on we want to make it consistent and decided to use function keyword in the component creation. 
Arrow function is a new way to create a function in es6. There are some differences between regular function and arrow function: https://dmitripavlutin.com/differences-between-arrow-and-regular-functions/

christiang2 replied:
Will do!

raymondz1 commented on cli/cmd/plugin/ui/web/tanzu-ui/src/shared/components/UnmanagedClusterInfo/UnmanagedClusterInfo.tsx:
```diff
@@ -0,0 +1,69 @@
+                        <div cds-layout="vertical m-r:xs">
+                            <label>Status</label>
+                            <div cds-layout="horizontal">
+                                {props.status == 'Running' ? (
```
```suggestion
                                {props.status === 'Running' ? (
```

Commit:
Made PR changes
 Changed getting started page from a const to a function

Commit:
Changed UnmanagedClusterInfo to reflect the three possible statuses of an unmanaged cluster

Commit:
Resolved Merge Conflicts
